[sreelakshmiharidos] iP

sreelakshmiharidos opened the pull request:
# Gilu – Your Personal Task Manager
> “Call it aim, not dream.” – Unknown  

**Gilu** is a chatbot that helps you manage your tasks efficiently. It is:  

- **Text-based**  
- **Easy to use**  
- **Lightweight and fast** 

### How to get started:
1. **Download** the latest release from [here](https://github.com/sreelakshmiharidos/ip/releases).  
2. **Run** the program by executing the JAR file.  
3. **Start managing** your tasks by adding, marking, and deleting them.  
4. **Enjoy** a clutter-free mind!  

### Features:
- [x] Manage tasks (To-Do, Deadlines, Events)  
- [x] Save tasks to a file  
- [ ] Implement reminders (coming soon)  
- [ ] Add support for recurring tasks (planned)  

### Code Example:
Here’s a snippet of Gilu’s `main` method:  
```java
public static void main(String[] args) {
        String storagePath = (args.length > 0) ? args[0] : DEFAULT_STORAGE_PATH;
        clearTestFileIfNeeded(args, storagePath);
        runGilu(storagePath);
    }

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Gave a new name to the chatbot and implemented welcome and exit messages.

Commit:
Updated the chatbot to read and echo user inputs and changed welcome and exit messages to make its personality unique.

Commit:
Updated the chatbot to maintain a task list to which user can add tasks and have it reproduced when asked.

Commit:
Added Task class and updated chatbot to have mark/unmark feature for the task list.

Commit:
Added support for Todo, Deadline, and Event tasks using inheritance.

Commit:
Implemented semi-automated testing.

Commit:
Implemented exception handling.

Commit:
Added delete feature to the chatbot.

Commit:
Replaced string-based command handling with enums for cleaner code.

BladerX11 commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+        if (input.startsWith("todo")) return TODO;
+        if (input.startsWith("deadline")) return DEADLINE;
+        if (input.startsWith("event")) return EVENT;
+        if (input.equalsIgnoreCase("bye")) return EXIT;
```
Should these conditional statements be put on a separate lines from the checks, wrapped in Egyptian style brackets?

BladerX11 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
+    }
```
Is it required for JavaDocs to be added here for descriptions of the public constructor parameters? I noticed the same issue in several constructors too.

BladerX11 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task {
```
Perhaps header comments can be added to this public class to provide a description of what it does? Several other classes seem to be missing it as well

nicholasdaijh replied:
All method names are not too long and not too short, great!

BladerX11 commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+                return false;
+            default:
+                throw new GiluException("Uh-oh! I didn’t get that. Try 'list', 'todo', 'deadline', 'event', or 'delete'.");
+        }
```
Should the indentation for case clauses be removed? Perhaps you could change the formatter settings in your IDE to have it done automatically for you?

BladerX11 commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+}
+
+public class Gilu {
+    private static final ArrayList<Task> tasks = new ArrayList<>();
```
Perhaps it would be suitable to use SCREAMING_SNAKE_CASE for naming as it is a constant?

BladerX11 commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    // Processes the user's command and throws GiluException for errors
+    private static boolean processCommand(String input) throws GiluException {
```
Should this be named like a boolean as it returns one?

nicholasdaijh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+public class Task {
```
Very good demonstration of using names to explain what the functions are doing, very descriptive

nicholasdaijh commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+}
+
+// Enum for command types
+enum Command {
```
Good job in fulfilling the requirement of using standard words (mark, unmark, delete, add) as method names

nicholasdaijh commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    // Displays the task list
+    private static void printTasks() {
```
printTasks() and printAddedTask() sound similar but do different things. Using displayTaskList() instead of printTasks() might improve clarity.

nicholasdaijh commented on src/main/java/Gilu.java:
```diff
@@ -0,0 +1,199 @@
+    LIST, MARK, UNMARK, DELETE, TODO, DEADLINE, EVENT, EXIT, UNKNOWN;
+
+    // Converts user input into an enum value
+    public static Command fromInput(String input) {
```
The method processes commands based on user input, so parseCommand() would better reflect its role, instead of fromInput()

nicholasdaijh reviewed (commented):
Good job overall!!

Commit:
Storage: Add task persistence to disk

Commit:
Add support for date and time handling in tasks

Commit:
Refactor code to improve OOP structure

Commit:
Organise classes into appropriate Java packages

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JUnit tests for addDeadline method

Commit:
Add JUnit tests for loadTasks method

Commit:
Update build.gradle to configure JAR creation

Commit:
Improve Javadoc documentation across classes

Commit:
Improve coding standard

Commit:
Add find feature to search tasks by keyword

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Refactor main method to follow SLAP

Commit:
Migrate Gilu from CLI to GUI-based app

Commit:
Add assertions to enforce assumptions

Various methods in the codebase rely on implicit assumptions,
such as valid indices for task retrieval, non-empty input for
task addition, and successful storage operations.

Lack of explicit checks means that violations of these
assumptions may lead to unexpected behavior or silent failures.

Let's introduce assertions to enforce these assumptions
throughout the codebase.

Adding assertions ensures that:
* Task indices are within valid bounds before access.
* Task descriptions and command inputs are non-empty.
* Storage operations execute without issues.

Assertions help catch logical errors early in development,
improving code reliability and simplifying debugging, while
remaining inactive in production unless explicitly enabled.

Commit:
Improve readability and maintainability

The codebase had several areas where readability and maintainability
could be improved. Issues included deep nesting, inline expressions,
and minor violations of coding best practices.

Improving code readability enhances ease of future modifications,
debugging, and overall comprehension while preserving original
functionality.

Let's apply the following refinements:
- Refactor method structures to avoid deep nesting and improve
logical flow.
- Use meaningful constants instead of magic literals for better
clarity.
- Ensure clear separation of concerns in constructors and
initialization blocks.
- Enhance method-level documentation to follow best practices.
- Improve readability of expressions by breaking complex conditions
into well-named boolean variables.

These changes make the codebase more maintainable, readable, and
aligned with standard coding practices.

sreelakshmiharidos committed:
Merge pull request #2 from sreelakshmiharidos/branch-A-Assertions

Add assertions to enforce assumptions

Commit:
Merge branch 'master' into branch-A-CodeQuality

sreelakshmiharidos committed:
Merge pull request #3 from sreelakshmiharidos/branch-A-CodeQuality

Commit:
Add general sorting for tasks by category and time

Commit:
Improve error handling

Commit:
Remove A-TextUiTesting

Commit:
Add representative screenshot

Commit:
Update README.md

Commit:
Update build.gradle

Commit:
Update build.gradle

Commit:
Add Checkstyle

Commit:
Improve code quality

Commit:
Improve code quality

Commit:
Update JUnit tests
